File: problem_solving/BOJ/BOJ_11376.cpp
#include <bits/stdc++.h>
using namespace std;
int am[2005];
int bm[2005];
vector<int> adj[2005];
int vis[2005];
int vcnt;
int dfs(int cur)
{
    if(vis[cur]==vcnt)return 0;
    vis[cur]=vcnt;
    for(int i=0;i<adj[cur].size();++i){
        int nxt=adj[cur][i];
        if(!bm[nxt]||(dfs(bm[nxt]))){
            bm[nxt]=cur;
            am[cur]=nxt;
            return 1;
        }
    }

    return 0;
}
int main()
{
    int n,m;
    scanf("%d%d",&n,&m);
    for(int i=1;i<=n;++i){
        int x;
        scanf("%d",&x);
        for(int j=0;j<x;++j){
            int y;
            scanf("%d",&y);
            adj[i].push_back(y);
            adj[i+n].push_back(y);
        }
    }
    int ans=0;
    for(int i=1;i<=(n<<1);++i){
        ++vcnt;
        ans+=dfs(i);
    }
    printf("%d\n",ans);
}

File: problem_solving/BOJ/BOJ_11376.py
import sys
sys.setrecursionlimit(10**6)
n,m=map(int,raw_input().split())
am=[-1]*(n<<1)
bm=[-1]*m
adj=[[] for i in xrange(n<<1)]
vcnt=0
vis=[0]*(n<<1)
for i in xrange(n):
	x=map(int,raw_input().split())
	for j in xrange(x[0]):
		adj[i].append(x[j+1]-1)
		adj[i+n].append(x[j+1]-1)
def dfs(cur):
	if(vis[cur]==vcnt):return 0
	vis[cur]=vcnt
	for nxt in adj[cur]:
		if(bm[nxt]==-1 or dfs(bm[nxt])):
			bm[nxt]=cur
			am[cur]=nxt
			return 1
	return 0
ans=0
for i in xrange(n<<1):
	vcnt+=1
	ans+=dfs(i)
print ans